File: go/cadreen/setup_sessions.go
package cadreen

import (
	"context"
	"fmt"
)

type SetupSessionCreateRequest struct {
	WorkspaceID string   `json:"workspace_id,omitempty"`
	Purpose     string   `json:"purpose,omitempty"`
	Constraints []string `json:"constraints,omitempty"`
}

type SetupSessionAddRequest struct {
	Connections []SetupConnection `json:"connections,omitempty"`
	Credentials []SetupCredential `json:"credentials,omitempty"`
	Memory      []SetupMemory     `json:"memory,omitempty"`
	Policies    []SetupPolicy     `json:"policies,omitempty"`
}

type SetupSessionApplyRequest struct {
	Confirm bool `json:"confirm"`
}

type SetupSession struct {
	ID           string            `json:"id"`
	Status       string            `json:"status"`
	Purpose      string            `json:"purpose,omitempty"`
	Constraints  []string          `json:"constraints,omitempty"`
	Connections  []SetupConnection `json:"connections"`
	Credentials  []SetupCredential `json:"credentials"`
	Memory       []SetupMemory     `json:"memory"`
	Policies     []SetupPolicy     `json:"policies"`
	Proposals    []SetupProposal   `json:"proposals,omitempty"`
	AppliedCount int               `json:"applied_count"`
	FailedCount  int               `json:"failed_count"`
	CreatedAt    string            `json:"created_at"`
	UpdatedAt    string            `json:"updated_at"`
	AppliedAt    string            `json:"applied_at,omitempty"`
}

type SetupSessionApplyResult struct {
	SessionID string       `json:"session_id"`
	Status    string       `json:"status"`
	Applied   int          `json:"applied"`
	Failed    int          `json:"failed"`
	Result    *SetupResult `json:"result,omitempty"`
}

type ListSetupSessionsResponse struct {
	Sessions []SetupSession `json:"sessions"`
}

func (c *Client) CreateSetupSession(ctx context.Context, req SetupSessionCreateRequest, opts ...RequestOption) (*SetupSession, error) {
	var result SetupSession
	if err := c.do(ctx, "POST", "/api/v1/cadreen/setup/sessions", req, &result, opts...); err != nil {
		return nil, fmt.Errorf("create setup session: %w", err)
	}
	return &result, nil
}

func (c *Client) ListSetupSessions(ctx context.Context, opts ...RequestOption) (*ListSetupSessionsResponse, error) {
	var result ListSetupSessionsResponse
	if err := c.do(ctx, "GET", "/api/v1/cadreen/setup/sessions", nil, &result, opts...); err != nil {
		return nil, fmt.Errorf("list setup sessions: %w", err)
	}
	return &result, nil
}

func (c *Client) GetSetupSession(ctx context.Context, id string, opts ...RequestOption) (*SetupSession, error) {
	var result SetupSession
	if err := c.do(ctx, "GET", "/api/v1/cadreen/setup/sessions/"+id, nil, &result, opts...); err != nil {
		return nil, fmt.Errorf("get setup session: %w", err)
	}
	return &result, nil
}

func (c *Client) AddToSetupSession(ctx context.Context, id string, req SetupSessionAddRequest, opts ...RequestOption) (*SetupSession, error) {
	var result SetupSession
	if err := c.do(ctx, "POST", "/api/v1/cadreen/setup/sessions/"+id, req, &result, opts...); err != nil {
		return nil, fmt.Errorf("add to setup session: %w", err)
	}
	return &result, nil
}

func (c *Client) ApplySetupSession(ctx context.Context, id string, req SetupSessionApplyRequest, opts ...RequestOption) (*SetupSessionApplyResult, error) {
	var result SetupSessionApplyResult
	if err := c.do(ctx, "POST", "/api/v1/cadreen/setup/sessions/"+id+"/apply", req, &result, opts...); err != nil {
		return nil, fmt.Errorf("apply setup session: %w", err)
	}
	return &result, nil
}

File: python/cadreen/resources/setup_sessions.py
from __future__ import annotations

from typing import Any

from ..client import HttpClient
from ..types import (
    SetupSession,
    SetupSessionApplyResult,
)


class SetupSessionsResource:
    def __init__(self, client: HttpClient) -> None:
        self._client = client

    async def create(
        self,
        *,
        purpose: str | None = None,
        constraints: list[str] | None = None,
        workspace_id: str | None = None,
    ) -> SetupSession:
        body: dict[str, Any] = {}
        if purpose is not None:
            body["purpose"] = purpose
        if constraints is not None:
            body["constraints"] = constraints
        if workspace_id is not None:
            body["workspace_id"] = workspace_id
        raw = await self._client.post("/api/v1/cadreen/setup/sessions", body)
        return self._parse_session(raw)

    async def list(self) -> list[SetupSession]:
        raw = await self._client.get("/api/v1/cadreen/setup/sessions")
        return [self._parse_session(s) for s in raw.get("sessions", [])]

    async def get(self, id: str) -> SetupSession:
        raw = await self._client.get(f"/api/v1/cadreen/setup/sessions/{id}")
        return self._parse_session(raw)

    async def add_resources(
        self,
        id: str,
        *,
        connections: list[dict[str, Any]] | None = None,
        credentials: list[dict[str, Any]] | None = None,
        memory: list[dict[str, Any]] | None = None,
        policies: list[dict[str, Any]] | None = None,
    ) -> SetupSession:
        body: dict[str, Any] = {}
        if connections is not None:
            body["connections"] = connections
        if credentials is not None:
            body["credentials"] = credentials
        if memory is not None:
            body["memory"] = memory
        if policies is not None:
            body["policies"] = policies
        raw = await self._client.post(f"/api/v1/cadreen/setup/sessions/{id}", body)
        return self._parse_session(raw)

    async def apply(self, id: str, *, confirm: bool = True) -> SetupSessionApplyResult:
        raw = await self._client.post(
            f"/api/v1/cadreen/setup/sessions/{id}/apply",
            {"confirm": confirm},
        )
        return SetupSessionApplyResult(
            session_id=raw["session_id"],
            status=raw["status"],
            applied=raw["applied"],
            failed=raw["failed"],
            result=raw.get("result"),
        )

    def _parse_session(self, raw: dict[str, Any]) -> SetupSession:
        return SetupSession(
            id=raw["id"],
            status=raw["status"],
            purpose=raw.get("purpose"),
            constraints=raw.get("constraints"),
            connections=raw.get("connections", []),
            credentials=raw.get("credentials", []),
            memory=raw.get("memory", []),
            policies=raw.get("policies", []),
            proposals=raw.get("proposals"),
            applied_count=raw.get("applied_count", 0),
            failed_count=raw.get("failed_count", 0),
            created_at=raw["created_at"],
            updated_at=raw["updated_at"],
            applied_at=raw.get("applied_at"),
        )
